### ISSUE TITLE
Web site not updated from Git

### ISSUE THREAD

**Issue by tbm** [ledger/ledger.github.io#46]
@peterkeen

I noticed the ledger-cli.org website is not updated when we make changes to this Git repo.

This is because Travis is not run anymore.

I see two options:

1) Enable Travis

2) Convert to GitHub Actions

I'd probably prefer 2) since that's the recommended solution these days, but I'm not sure how the S3 credentials are stored/access from GitHub Actions.

**Comment by peterkeen** [comment]
@tbm Thanks for the heads up, I'll get it working again.

**Comment by afh** [comment]
@peterkeen is #33 something you could possibly address too?

**Comment by peterkeen** [comment]
I'm having an issue building the PDFs. The system that worked on Travis is blowing up on Unicode literals and I don't know enough about tex to make progress. 

Can anyone take a look at the recent Action builds and give me a pointer?

**Comment by afh** [comment]
@peterkeen I see texinfo versino 6.8 is being used, is there a possibility to use a newer version, e.g. 7.0.2?

**Comment by afh** [comment]
Do we know what the locale on the system is? Maybe it needs the `.UTF-8` suffix, e.g. `LANG=en_US.UTF-8`…

**Comment by afh** [comment]
FYI: I'm able to build the documentation locally on macOS with nixpkgs using texinfo 6.8.

[help-texinfo](https://lists.gnu.org/archive/html/help-texinfo/2021-01/msg00021.html) suggests to:
``
@iftex
\DeclareUnicodeCharacter{00fa}{\'u}
@end iftex
```

**Comment by tbm** [comment]
What's the status of this?

**Comment by peterkeen** [comment]
I had a bunch of personal stuff come up and haven't had time or energy to figure it out. If someone wants to take a crack at it feel free.

**Comment by peterkeen** [comment]
I just merged @afh 's PR, we'll see if that fixes it.

**Comment by afh** [comment]
🤞

**Comment by tbm** [comment]
The build shows as successful but https://www.ledger-cli.org/download.html still shows the old Gentoo version.

**Comment by afh** [comment]
Looking at one of the last [builds on travis-ci.org](https://travis-ci.org/github/ledger/ledger-website) I see that there is a `Preparing deploy` and `Deploying application` phase, which does not show up in the [GitHub Actions](https://github.com/ledger/ledger-website/actions/runs/4314820113/jobs/7528358811). Looking at the `.travis.yml` file it features a [`deploy` section](https://github.com/ledger/ledger-website/blob/master/.travis.yml#L21-L28) and there seems to be no equivalent in the [`build_and_deploy.yml`](https://github.com/ledger/ledger-website/blob/master/.github/workflows/build_and_deploy.yml) that drives the GitHub Actions.

Inspecting the [**Secrets and variables** for Actions](https://github.com/ledger/ledger-website/settings/secrets/actions) in this repositories settings I do not see any AWS or S3 related credentials. @peterkeen would you be able to add [those](https://github.com/ledger/ledger-website/blob/master/.travis.yml#L23-L26)?

After that I'm happy to look into integrating or adding an action (possibly from the [GitHub Marketplace](https://github.com/marketplace?category=&query=s3+sort%3Apopularity-desc&type=&verification=)) to deploy to S3.

**Comment by simonmichael** [comment]
If all else fails, there is also the older Github Pages setup, which was simple and worked.

**Comment by peterkeen** [comment]
@afh Done! My plan was to get the build working and then get the S3 deploy happening but it obviously didn't turn out that way.

**Comment by afh** [comment]
Thanks, @peterkeen! Happy to help out on getting the deployment back up, but also do not want to get in your way if you plan on working on this. Would you like to take lead on this or shall I?

Good to know @simonmichael, thanks!

**Comment by peterkeen** [comment]
@afh Please feel free to take point on getting it back up and running!

**Comment by afh** [comment]
Thanks, @peterkeen, the result of me having taken point is PR #53; looking forward to your review and comments.

**Comment by tbm** [comment]
I applied @afh's changes (thanks!) but as far as I can tell the website has not updated.

**Comment by afh** [comment]
Let me have a closer look…

**Comment by tbm** [comment]
The good news: the web site has updated now.

The web news. Docs are broken:

> 11:31 < hamess> getting AccessDenied error on all /doc pages: https://www.ledger-cli.org/doc/ledger3.html

**Comment by afh** [comment]
Thanks for the update, @tbm. I can confirm you observations. I assume the denied access is related to missing change in the S3 configuration for the bucket, since we changed the URLs to the docs.

For example the old [Ledger 2.6 documentation](https://www.ledger-cli.org/2.6/doc/ledger.html) residing at the old URL (which has been moved to a static site and is not longer dynamically rendered from the `maint` branch) is accessible, but broken; likely due to the fact, that it has been processed by a Markdown processor or middleman layout. I'll look into it.

@tbm or @jwiegley do you happen to have access to the S3 bucket powering ledger-cli.org by any chance?

**Comment by peterkeen** [comment]
See https://github.com/ledger/ledger-website/issues/59

**Comment by afh** [comment]
Making a little bit of progress (see https://ledger.github.io/ledger-website for current state). I'll continue working on the migration to GitHub Pages CET tomorrow.

**Comment by afh** [comment]
For testing purposes I forked ledger/ledger-website to ledger/ledger.github.io, this allowed to deploy the website to https://ledger.github.io and not worry about any absolute/relative paths, that change when deploying to a GitHub page within the ledger/ledger-website repo.

I kindly request that folks check out https://ledger.github.io (including the [deprecated documentation](https://ledger.github.io/2.6/doc/ledger.html) for ledger 2.x).

If folks are okay with the contents and how things look, we can talk about whether we want to rename the ledger-website repo to ledger.github.io (possibly breaking URLs) or archive ledger-website with a notice that it has been superseded by the ledger.github.io repo.

Looking forward to your feedback and if you think that things should be done differently I'd love to hear your ideas.

**Comment by simonmichael** [comment]
Looks nice Alexis. All links I tried worked. I didn't find a link to the 2.6 docs, is that intended ?

My 2c re the domain, unless John feels like a change I think it would be a shame to needlessly break all existing ledger-cli.org <http://ledger-cli.org/> links and tie the site to GIthub's domain space for no good reason.

**Comment by peterkeen** [comment]
@simonmichael I think @afh was referring to renaming the _repo_, not the website. [GitHub preserves redirects indefinitely](https://github.com/orgs/community/discussions/22669) so I don't think there is any reason to not rename.

**Comment by simonmichael** [comment]
Oops! Carry on!

**Comment by afh** [comment]
Thanks for checking all the links, @simonmichael. Yes, that the 2.6 docs are not linked from anywhere is intentional. Here's my thinking around it: If someone has a visits page directly, e.g. via a bookmark, it is still served with an added deprecation notice, yet it "should" not be found otherwise as to avoid confusion, given that 2.6 has been deprecated for quite some time.
I'm happy to include a link with the appropriate deprecation version if people think there is value in it.

And yes, I was "referring to renaming the _repo_, not the website" as @peterkeen (ty ty) correctly clarified. In my mind ledger.github.io hosts the content and ledger-cli.org "points" to it. I'll look into custom domain names for GitHub pages today.

@peterkeen, do you have an idea how we could migrate the Lambda@Edge function to a different technology to maintain the ledger-cli.org subdomains? I'll look into  creating an AWS account and possibly taking over that Lamda@Edge function. Though I'd prefer a solution involving technology that isn't gated by an individual account.

**Comment by afh** [comment]
Quick update on the migration to GitHub pages:

* It is still very much work in progress and I'm testing things out
* I've configured a custom domain name in the settings of the ledger.github.io repo and wait for the DNS check to complete
* I expect it to fail as the ledger-cli.org DNS isn't configured yet for use with GitHub Pages. Someone with access to the DNS (is that you @jwiegley?) would need to configure the DNS properly (see [**Managing a custom domain for your GitHub Pages site**](https://docs.github.com/en/pages/configuring-a-custom-domain-for-your-github-pages-site/managing-a-custom-domain-for-your-github-pages-site#about-custom-domain-configuration) for details).
* Here's a list of todos that will need to happen to officially migrate:
  - [ ] Rename the ledger/ledger-website repo to ledger/ledger.github.io (@afh)
  - [ ] Rewrite GitHub build-and-deploy action to deploy to GitHub Pages (see [`build-and-deploy.yml` below](#build-and-deploy) for example)
  - [ ] Configure GitHub Pages in ledger/ledger.github.io (@afh)
     - Deploy from GitHub Actions
     - Set custom domain to www.ledger-cli.org
     - Check GitHub Pages deployment branches in [**Environments**](https://github.com/ledger/ledger.github.io/settings/environments)
  - [ ] Configure DNS for ledger-cli.org (at least apex domain and www subdomain;  [see GitHub docs](https://docs.github.com/en/pages/configuring-a-custom-domain-for-your-github-pages-site/managing-a-custom-domain-for-your-github-pages-site#configuring-an-apex-domain) for details) (@jwiegley?)
  - [ ] Come up with an alternative solution to Lambda@Edge to handle git, wiki, list subdomains (@afh, @tbm, @peterkeen)
  
<details><summary><a name="build-and-deploy"><code>build-and-deploy.yml</code></a></summary>

```
name: Build and Deploy

on:
  workflow_dispatch:
  push:
    branches: [ "master" ]
  pull_request:
    branches: [ "master" ]

permissions:
  contents: read

jobs:
  build_and_deploy:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v3
    - name: Install dependencies
      run: |
        sudo apt-get update;
        sudo apt-get install --yes \
          texinfo groff \
          texlive-base texlive-fonts-recommended texlive-plain-generic
    - name: Set up Ruby
      uses: ruby/setup-ruby@v1
      with:
        ruby-version: 2.7
        bundler-cache: true # runs 'bundle install' and caches installed gems automatically
    - name: Build website
      run: bundle exec middleman build
    - name: Build documentation
      run: make docs
    - name: Deploy
      if: ${{ github.ref == 'refs/heads/master' }}
      env:
        AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
        AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
      uses: reggionick/s3-deploy@v3
      with:
        folder: build
        bucket: ${{ vars.AWS_S3_BUCKET }}
        bucket-region: ${{ vars.AWS_S3_BUCKET_REGION }}
        delete-removed: true
        no-cache: true
        private: true
        filesToInclude: ".*/*,*/*,**"
```

</details>

**Comment by afh** [comment]
As expected the GitHub Pages custom DNS check failed with:
> Both www.ledger-cli.org and its alternate name are improperly configured
Your site's DNS settings are using a custom subdomain, www.ledger-cli.org, that is set up as an A record. We recommend you change this to a CNAME record pointing to ledger.github.io. For more information, see [documentation](https://docs.github.com/articles/setting-up-a-custom-domain-with-github-pages/) (InvalidARecordError).

**Comment by afh** [comment]
What are people's opinion on using http://redirect.center (docs: http://redirect.name) to redirect the (git|list|wiki).ledger-cli.org subdomains to the respective URLs and replace the current Lambda@Edge (see [ledger's about page](https://www.ledger-cli.org/about.html) for details).

**Comment by afh** [comment]
I think cloudflare redirects are better suited than the previously mentioned redirect.center aproach. I've created a cloudflare account (happy to share credentials with other maintainers) and added a "website" for ledger-cli.org in cloudflare. In order to move forward with the migration the current DNS for ledger-cli.org needs to be updated (see [cloudflare documentation](https://developers.cloudflare.com/dns/zone-setups/full-setup/setup/#update-your-nameservers) for details). Once those changes have propagated through the net I'll wrap up the [other tasks](https://github.com/ledger/ledger-website/issues/46#issuecomment-1463394782).

**Comment by peterkeen** [comment]
That sounds fine to me. The owner of the domain (I think @jwiegley ?) will need to do the registrar work.

**Comment by jwiegley** [comment]
Sure, just let me know what you need done.

**Comment by afh** [comment]
Thanks @jwiegley. Here is what needs to get done:

1. Log in to the ledger-cli.org domain registrar account
2. Remove the following nameservers for the ledger-cli.org domain:
   * ns-1183.awsdns-19.org
   * ns-1994.awsdns-57.co.uk
   * ns-422.awsdns-52.com
   * ns-974.awsdns-57.net
3. Add Cloudflare's nameservers for the ledger-cli.org domain:
   * andronicus.ns.cloudflare.com
   * deb.ns.cloudflare.com
4. Write a quick message here once the above steps have been completed as the nameserver updates should take about 24 hours to complete after which I'll take care of the remaining tasks to migrate ledger-cli.org to GitHub Pages.
5. If DNSSEC had to be disabled when replacing the nameservers, please let me know, so that I can enable it once the nameserver changes have propagated through DNS (usually 24 hours) and I will have to ask you to add the new DS record once DNSEC is re-enabled (for details see [Cloudflare's article on DNSSEC](https://developers.cloudflare.com/dns/additional-options/dnssec/))
6. Let me know who else (email and PGP public key) should have access to the Cloudflare account, so that I can share credentials with them, so that in case things go awry more people have access to the infrastructure.

**Comment by afh** [comment]
ℹ️  I've renamed the repo to ledger.github.io and am working on improving how the documentation is build (see #63 and [`afh/ledger@web-docs`](https://github.com/ledger/ledger/compare/master...afh:ledger:web-docs) for details).

**Comment by jwiegley** [comment]
@afh I followed the steps above and switched the nameservers. DNSSEC did not come up.
<img width="1236" alt="ledger-cli org Domain Management  Name com 2023-03-23 14-35-14" src="https://user-images.githubusercontent.com/8460/227369173-c93abf58-7723-4c7b-b3d3-c5677600a202.png">

**Comment by afh** [comment]
Thank you very much, @jwiegley, I'll work on the [remaining tasks](https://github.com/ledger/ledger.github.io/issues/46#issuecomment-1463394782) to complete the migration and may come back to you with polite requests for domain registrar related tasks that need doing (e.g. DS record).

**Comment by afh** [comment]
https://ledger-cli.org is now working for me (including www-subdomain and ledger.github.io redirects) :tada:

Due to the changes in DNS please allow for a grace period of 24–48 hours before reporting issues. Currently the `git`, `wiki`, and `list` subdomains only resolve when using clouldflare's DNS (`8.8.8.8`), yet that should be fully propagated throughout the network after the grace period, i.e. after the weekend.

**Comment by afh** [comment]
It would be nice if folks could chime in from different parts of the globe with a quick:
`Works for me from :YOUR_RESIDENCE_COUNTRY_FLAG_EMOJI::+1:`, e.g. Works for me from :eu::+1:

**Comment by tbm** [comment]
Works for me from :philippines:. Thanks

**Comment by afh** [comment]
Closing as fixed ✅
